src/common/Journald.h

```diff
@@ -0,0 +1,70 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef __CEPH_LOG_JOURNALD_H
+#define __CEPH_LOG_JOURNALD_H
+
+#include <memory>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+struct LogEntry;
+
+namespace ceph {
+
+namespace logging {
+
+namespace detail {
+class EntryEncoder;
+class LogEntryEncoder;
+
+class JournaldClient {
+ public:
+  JournaldClient();
+  ~JournaldClient();
+  void send();
+  struct msghdr m_msghdr;
+ private:
+  int fd;
+};
+}
+
+class Entry;
+class SubsystemMap;
+
+class Journald {
+ public:
+  Journald(const SubsystemMap *s);
+  ~Journald();
+
+  void log_entry(const Entry &e);
+
+  using Ref = std::shared_ptr<Journald>;
+
+ private:
+  detail::JournaldClient client;
+
+  std::unique_ptr<detail::EntryEncoder> m_entry_encoder;
+
+  const SubsystemMap * m_subs;
+};
+
+class JournaldClusterLog {
+ public:
+  JournaldClusterLog();
+  ~JournaldClusterLog();
+
+  void log_log_entry(const struct LogEntry &le);
+
+  using Ref = std::shared_ptr<JournaldClusterLog>;
```

Comment:
why? the only user of this type is `LogMonitor::log_channel_info::get_journald()`. which in turn is called by `LogMonitor::update_from_paxos()` which populate the logging messages to the enabled cluster loggers. but `LogMonitor` does not own an instance of `JournaldClusterLog`. so returning a reference to the `journald` is good enough, i think.

Reply:
I just copied some code from Graylog. Following your comments now.